lib/jira/util/index.js

```diff
@@ -1,19 +1,45 @@
-const jiraIssueReferenceRegex = /\[([A-Z]+-[0-9]+)\](?!\()/g
+const { markdown } = require('markdown')
 
 module.exports = function (jiraClient) {
   function addJiraIssueLinks (text, issues) {
+    const referenceRegex = /\[([A-Z]+-[0-9]+)\](?!\()/g
     const issueMap = issues.reduce((issueMap, issue) => ({
       ...issueMap,
       [issue.key]: issue
     }), {})
 
-    return text.replace(jiraIssueReferenceRegex, (match, issueKey) => {
-      if (!issueMap[issueKey]) {
-        return match
+    const links = []
+
+    // Attempt to parse our existing issue text, pulling out any existing reference links.
+    const [, { references }] = markdown.parse(text)
+    const keys = references ? Object.keys(references) : []
+
+    // Parse the text up to a maximum amount of characters.
+    while (referenceRegex.lastIndex < 1000) {
+      const match = referenceRegex.exec(text)
+
+      if (!match) {
+        break
+      }
+
+      const [, key] = match
+
+      // If we already have a reference link, or the issue is not valid, skip it.
+      if (keys.includes(key.toLowerCase()) || !issueMap[key]) {
+        continue
       }
-      const link = `${jiraClient.baseURL}/browse/${issueKey}`
-      return `[[${issueKey}]](${link})`
-    })
+
+      const link = `${jiraClient.baseURL}/browse/${key}`
+      const reference = `[${key}]: ${link}`
+
+      if (text.includes(reference)) {
+        continue
+      }
+
+      links.push(reference)
+    }
+
+    return links.length ? [text, links].join('\n\n') : text
```

Comment:
One minor suggestion here to improve how it renders if you place multiple links in the same comment. Without this it was displaying multiple issues like this:

![image](https://user-images.githubusercontent.com/13207348/52389697-c7472180-2a62-11e9-9cbe-59a0f4fee876.png)


```suggestion
    return links.length ? [text, links.join('\n'].join('\n\n') : text
```

Reply:
ah, thats embarrassing, sorry about that! just pushed a fix + coverage